##########################################
Date: Jan 25, 2021 3:00 PM
From: Sushma Natarajan
Subject: WLDM Integration
Source: Broadcom
License: GPLV2
Upstream-Status: Pending
Signed-off-by: Sushma Natarajan <sushma.natarajan@broadcom.com>
##########################################

diff --git a/source/TR-181/sbapi/cosa_wifi_apis.c b/source/TR-181/sbapi/cosa_wifi_apis.c
index 0687c173..8d612545 100755
--- a/source/TR-181/sbapi/cosa_wifi_apis.c
+++ b/source/TR-181/sbapi/cosa_wifi_apis.c
@@ -10204,6 +10205,12 @@ fprintf(stderr, "----# %s %d gRadioRestartRequest=%d %d \n", __func__, __LINE__,
 			gRadioRestartRequest[1]=FALSE;
 		}
 		//<<
+#if 1 /* CMWIFI_RDKB, disable restart in CosaDmlWiFiRadioSetCfg. */
+		if (wlanRestart) {
+			wifiDbgPrintf("### %s: ignore wlanRestart for instance %ld ###\n", __FUNCTION__, pCfg->InstanceNumber - 1);
+			wlanRestart = FALSE;
+		}
+#endif /* CMWIFI_RDKB */
 		if(wlanRestart == TRUE)
 		{
             memcpy(&sWiFiDmlRadioStoredCfg[pCfg->InstanceNumber-1], pCfg, sizeof(COSA_DML_WIFI_RADIO_CFG));
@@ -10232,6 +10239,11 @@ fprintf(stderr, "----# %s %d gRadioRestartRequest=%d %d \n", __func__, __LINE__,
 		    system(cmd);
 		}
 #endif
+
+#if 1 /* CMWIFI_RDKB, call wifi_apply in CosaDmlWiFiRadioSetCfg. */
+        fprintf(stderr, "%s: calling wifi_apply()...\n", __func__);
+        wifi_apply();
+#endif /* CMWIFI_RDKB */
     }
    
     return ANSC_STATUS_SUCCESS;
@@ -10958,6 +10970,11 @@ wifiDbgPrintf("%s\n",__FUNCTION__);
     else
         returnStatus = ANSC_STATUS_FAILURE;
 
+#if 1 /* CMWIFI_RDKB, call wifi_apply in CosaDmlWiFiSsidDelEntry. */
+    fprintf(stderr, "%s: calling wifi_apply()\n", __func__);
+    wifi_apply();
+#endif /* CMWIFI_RDKB */
+
     return returnStatus;
 #endif
 }
@@ -14222,6 +14239,11 @@ wifiDbgPrintf("%s\n",__FUNCTION__);
 #endif /* _ENABLE_BAND_STEERING_ */
 	}
 
+#if 1 /* CMWIFI_RDKB, call wifi_apply in CosaDmlWiFiApMfSetCfg. */
+    fprintf(stderr, "%s: calling wifi_apply()...\n", __func__);
+    wifi_apply();
+#endif /* CMWIFI_RDKB */
+
     return ANSC_STATUS_SUCCESS;
 }
 
