Date: May 24, 2020 1:41 PM
From: Iac5629e0e058f5814b90781dc10d6951b43fc4d1 May 24, 2020 1:41 PM
Subject: RDKB-29192
Source: COMCAST
License: Apache-2.0
Upstream-Status: Pending
Signed-off-by: kjothi774 <Krishnaveni_Jothimathavan@comcast.com>
Index: git/src/privilege.c
===================================================================
--- git.orig/src/privilege.c
+++ git/src/privilege.c
@@ -1,6 +1,24 @@
 #include "privilege.h"
+#include "parodus_log.h"
+#include "cap.h"
+#include "syscfg/syscfg.h"
+static cap_user appcaps;
 
 void drop_root_privilege()
 {
-
+  char buf[8] = {'\0'};
+  appcaps.caps = NULL;
+  appcaps.user_name = NULL;
+  syscfg_init();
+  syscfg_get( NULL, "NonRootSupport", buf, sizeof(buf));
+  if( buf != NULL )
+  {
+      if (strncmp(buf, "true", strlen("true")) == 0) {
+          init_capability();
+          drop_root_caps(&appcaps);
+          update_process_caps(&appcaps);
+          read_capability(&appcaps);
+      }
+      ParodusInfo("RFC Setting for NonRootSupport is: %s\n",buf);
+  }
 }
