Date: May 24, 2020 1:41 PM
From: Iac5629e0e058f5814b90781dc10d6951b43fc4d1 May 24, 2020 1:41 PM
Subject: RDKB-29192
Source: COMCAST
License: Apache-2.0
Upstream-Status: Pending
Signed-off-by: kjothi774 <Krishnaveni_Jothimathavan@comcast.com>
Index: git/src/privilege.c
===================================================================
--- git.orig/src/privilege.c
+++ git/src/privilege.c
@@ -1,6 +1,24 @@
 #include "privilege.h"
+#include "parodus_log.h"
+#include "cap.h"
+static cap_user appcaps;
 
 void drop_root_privilege()
 {
-
+  appcaps.caps = NULL;
+  appcaps.user_name = NULL;
+  bool ret = false;
+  ret = isBlocklisted();
+  if(ret)
+  {
+       ParodusInfo("NonRoot feature is disabled\n");
+  }
+  else
+  {
+       ParodusInfo("NonRoot feature is enabled, dropping root privileges for parodus process\n");
+       init_capability();
+       drop_root_caps(&appcaps);
+       update_process_caps(&appcaps);
+       read_capability(&appcaps);
+  }
 }
