Date: August 25, 2020 1:41 PM
From: Iac5629e0e058f5814b90781dc10d6951b43fc4d1 August 25, 2020 1:41 PM
Subject: RDKB-29192
Source: COMCAST
License: Apache-2.0
Upstream-Status: Pending
Signed-off-by: mkrish802 <Manoj_Krishnan@comcast.com>

Index: git/source/app/privilege.c
===================================================================
--- git.orig/source/app/privilege.c
+++ git/source/app/privilege.c
@@ -1,6 +1,25 @@
 #include "privilege.h"
+#include "webpa_adapter.h"
+#include "cap.h"
+
+static cap_user appcaps;
 
 void drop_root_privilege()
 {
-
+   appcaps.caps = NULL;
+   appcaps.user_name = NULL;
+   bool ret = false;
+   ret = isBlocklisted();
+   if(ret)
+   {
+       WalInfo("NonRoot feature is disabled\n");
+   }
+   else
+   {
+       WalInfo("NonRoot feature is enabled, dropping root privileges for webpa process\n");
+       init_capability();
+       drop_root_caps(&appcaps);
+       update_process_caps(&appcaps);
+       read_capability(&appcaps);
+   }
 }
