Date: May 10, 2016 8:39 AM

From: 59f650bc8f22182e2e1752e50826e910b0c8f55f Jul 29, 2016 4:32 AM

Subject: [PATCH] RDKB-5474: Handled APS1 firmwares supporting V2 flag as well as not.

	Initial support for RSSI.
     
Source: COMCAST


Upstream-Status: Pending

Signed-off-by: Ramki Ananthakrishnan <ramki_ananthakrishnan@cable.comcast.com>
---

diff -Nur a/upnp/src/ssdp/ssdp_server.c b/upnp/src/ssdp/ssdp_server.c
--- a/upnp/src/ssdp/ssdp_server.c	2016-05-09 23:54:28.518346756 +0000
+++ b/upnp/src/ssdp/ssdp_server.c	2016-05-09 23:55:55.290326113 +0000
@@ -516,26 +516,37 @@
 		}
 	}
 	//patched by cisco to support wecb device
-	if (strstr(cmd, "hnap:WiFiExtender") != NULL) 
-	{
-		memset(Evt->DeviceType, 0, sizeof(Evt->DeviceType));
-		if ((!strncmp(cmd, "hnap:WiFiExtender", strlen("hnap:WiFiExtender") + 1))) 
-		{
-			strncpy(Evt->DeviceType, "hnap:WiFiExtender", sizeof(Evt->DeviceType) - 1);
-			CommandFound = 1;
-			return 0;
-		}
-		else if(!strncmp(cmd, "hnap:WiFiExtenderSynced", strlen("hnap:WiFiExtenderSynced") + 1))
-		{
-			strncpy(Evt->DeviceType, "hnap:WiFiExtenderSynced", sizeof(Evt->DeviceType) - 1);
-			return 0;
-		}
-		else
-		{
-			printf("error packet:%s\n", cmd);
-			return -1;
-		}
-	}
+        if (strstr(cmd, "hnap:WiFiExtender") != NULL)
+        {
+                memset(Evt->DeviceType, 0, sizeof(Evt->DeviceType));
+                if (!strncmp(cmd, "hnap:WiFiExtender", strlen("hnap:WiFiExtender") + 1))
+                {
+                        strncpy(Evt->DeviceType, "hnap:WiFiExtender", sizeof(Evt->DeviceType) - 1);
+                        CommandFound = 1;
+                        return 0;
+                }
+                else if (!strncmp(cmd, "hnap:WiFiExtenderV2", strlen("hnap:WiFiExtenderV2") + 1))
+                {
+                        strncpy(Evt->DeviceType, "hnap:WiFiExtenderV2", sizeof(Evt->DeviceType) - 1);
+                        CommandFound = 1;
+                        return 0;
+                }
+                else if(!strncmp(cmd, "hnap:WiFiExtenderSynced", strlen("hnap:WiFiExtenderSynced") + 1))
+                {
+                        strncpy(Evt->DeviceType, "hnap:WiFiExtenderSynced", sizeof(Evt->DeviceType) - 1);
+                        return 0;
+                }
+                else if(!strncmp(cmd, "hnap:WiFiExtenderSyncedV2", strlen("hnap:WiFiExtenderSyncedV2") + 1))
+                {
+                        strncpy(Evt->DeviceType, "hnap:WiFiExtenderSyncedV2", sizeof(Evt->DeviceType) - 1);
+                        return 0;
+                }
+                else
+                {
+                        printf("error packet:%s\n", cmd);
+                        return -1;
+                }
+        }
 
 	if ((TempPtr = strstr(cmd, "::upnp:rootdevice")) != NULL) {
 		/* Everything before "::upnp::rootdevice" is the UDN. */
